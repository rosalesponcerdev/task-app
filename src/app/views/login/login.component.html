<div class="container">
  <div class="row align-items-center vh-100">
    <div class="col">
      <div class="w-50 mx-auto">
        <fieldset class="border border-2 px-3 pt-3 pb-5">
          <legend class="float-none m-0">LOGIN</legend>

          <span
            class="text-danger fw-bold"
            [ngClass]="{
              'v-hidden': !errorMessage
            }">
            {{ errorMessage?.message || '|' }}
          </span>

          <form class="row align-items-center" [formGroup]="loginPresenter.loginForm" (ngSubmit)="sendFormHandler()">
            <div class="col-8">
              <div class="mb-3">
                <label for="usernameinput" class="form-label">Usuario</label>
                <input
                  [ngClass]="{
                    'is-invalid': loginPresenter.loginForm.controls['username']!.errors?.['credentialFalse']
                  }"
                  formControlName="username"
                  type="text"
                  class="form-control"
                  id="usernameinput"
                  autocomplete="off" />
              </div>

              <div>
                <label for="passinput" class="form-label">Contrase√±a</label>
                <input
                  [ngClass]="{
                  'is-invalid': loginPresenter.loginForm.controls['password']!.errors?.['credentialFalse']
                }"
                  formControlName="password"
                  type="password"
                  class="form-control"
                  id="passinput" />
              </div>
            </div>
            <div class="col-4 text-center">
              <button [disabled]="!loginPresenter.loginForm.valid" type="submit" class="btn btn-primary">
                @if (loading) {
                <span class="spinner-border spinner-border-sm" aria-hidden="true"></span>
                } @else{
                <span>Ingresar</span>
                }
              </button>
            </div>
          </form>
        </fieldset>
      </div>
    </div>
  </div>
</div>
