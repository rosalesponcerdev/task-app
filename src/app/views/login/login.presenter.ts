import { Injectable, OnDestroy } from '@angular/core';
import { FormControl, FormGroup, Validators } from '@angular/forms';
import { LoginForm } from '../../interface/login-form.interface';
import { Observable, Subject, takeUntil } from 'rxjs';

@Injectable()
export class LoginPresenterService implements OnDestroy {
  loginForm: FormGroup<{
    username: FormControl<string>;
    password: FormControl<string>;
  }>;

  sendFormHandler$: Observable<LoginForm>;
  private _sendFormHandler: Subject<LoginForm>;

  private _destroy = new Subject();

  constructor() {
    this._sendFormHandler = new Subject<LoginForm>();
    this.sendFormHandler$ = this._sendFormHandler.asObservable().pipe(takeUntil(this._destroy));

    this.loginForm = this._buildLoginForm();
  }

  ngOnDestroy(): void {
    this._destroy.next(true);
    this._destroy.complete();
  }

  sendFormHandler() {
    if (!this.loginForm.valid) return;

    this.loginForm.disable();

    this._sendFormHandler.next(this.getFormValue());
  }

  getFormValue() {
    return this.loginForm.value as LoginForm;
  }

  setCredentialError(errorMessage: string) {
    this.loginForm.reset();
    this.loginForm.enable();

    this.loginForm.controls.username.setErrors(
      {
        credentialFalse: errorMessage,
      },
      {
        emitEvent: true,
      }
    );

    this.loginForm.controls.password.setErrors(
      {
        credentialFalse: errorMessage,
      },
      {
        emitEvent: true,
      }
    );
  }

  private _buildLoginForm() {
    return new FormGroup({
      username: new FormControl<string>('', {
        validators: [Validators.required],
        nonNullable: true,
      }),
      password: new FormControl<string>('', {
        validators: [Validators.required],
        nonNullable: true,
      }),
    });
  }
}
