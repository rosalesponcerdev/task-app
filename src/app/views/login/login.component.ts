import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output, SimpleChanges } from '@angular/core';

import { LoginPresenterService } from './login.presenter';
import { CommonModule } from '@angular/common';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { LoginForm } from '../../interface/login-form.interface';

@Component({
  selector: 'app-login-ui',
  standalone: true,
  imports: [CommonModule, ReactiveFormsModule, FormsModule],
  templateUrl: './login.component.html',
  providers: [LoginPresenterService],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class LoginComponent {
  @Input() errorMessage?: { message: string } = { message: '' };
  @Input() loading: boolean | null = false;

  @Output() sendForm = new EventEmitter<LoginForm>();

  constructor(public loginPresenter: LoginPresenterService) {
    this.loginPresenter.sendFormHandler$.subscribe((formValue) => {
      this.sendForm.emit(formValue);
    });
  }

  ngOnInit(): void {}

  ngOnChanges(changes: SimpleChanges): void {
    if (changes['errorMessage'] && !changes['errorMessage'].firstChange) {
      this.loginPresenter.setCredentialError(changes['errorMessage'].currentValue?.message);
    }
  }

  sendFormHandler() {
    this.loginPresenter.sendFormHandler();
  }
}
