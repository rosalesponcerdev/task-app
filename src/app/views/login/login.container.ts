import { Component, OnInit } from '@angular/core';
import { LoginComponent } from './login.component';
import { LoginForm } from '../../interface/login-form.interface';
import { Store } from '@ngrx/store';
import { login } from '../../store/session/session.actions';
import { Observable } from 'rxjs';
import { selectIsLoading, selectSessionError } from '../../store/session/session.selectors';
import { CommonModule } from '@angular/common';

@Component({
  selector: 'app-login-container',
  standalone: true,
  template: `
    <app-login-ui [errorMessage]="error$ | async" [loading]="loading$ | async" (sendForm)="sendFormHandler($event)" />
  `,
  imports: [LoginComponent, CommonModule],
})
export class LoginContainer implements OnInit {
  error$: Observable<any>;
  loading$: Observable<boolean>;

  constructor(private _store: Store) {
    this.error$ = this._store.select(selectSessionError);
    this.loading$ = this._store.select(selectIsLoading);
  }

  ngOnInit() {}

  sendFormHandler(loginForm: LoginForm) {
    this._store.dispatch(login(loginForm));
  }
}
