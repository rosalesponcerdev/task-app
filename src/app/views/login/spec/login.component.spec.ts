import { ComponentFixture, TestBed } from '@angular/core/testing';

import { LoginComponent } from '../login.component';
import { CommonModule } from '@angular/common';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { SimpleChanges } from '@angular/core';

describe('LoginComponent', () => {
  let component: LoginComponent;
  let fixture: ComponentFixture<LoginComponent>;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [LoginComponent],
    }).compileComponents();

    fixture = TestBed.createComponent(LoginComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
    expect(fixture.nativeElement).toMatchSnapshot();
  });

  it('Loggin Success', () => {
    expect(fixture.nativeElement).toMatchSnapshot();

    jest.spyOn(component.sendForm, 'emit');

    let loginButton = (fixture.nativeElement as HTMLElement).querySelector('button')!;
    expect(loginButton).toBeTruthy();
    expect(loginButton.disabled).toBeTruthy();

    const [usernameInput, passwordInput] = Array.from((fixture.nativeElement as HTMLElement).querySelectorAll('input')!);
    expect(usernameInput).toBeTruthy();
    expect(passwordInput).toBeTruthy();

    component.loginPresenter.loginForm.patchValue({
      password: 'test01',
      username: 'test01',
    });

    component.loginPresenter.loginForm.markAsTouched();
    component.loginPresenter.loginForm.updateValueAndValidity();

    fixture.detectChanges();

    expect(fixture.nativeElement).toMatchSnapshot();

    expect(component.loginPresenter.loginForm.valid).toBeTruthy();

    component.sendFormHandler();

    expect(component.sendForm.emit).toHaveBeenCalled();
  });

  it('Loggin Error', () => {
    expect(fixture.nativeElement).toMatchSnapshot();

    component.errorMessage = {
      message: 'message',
    };
    component.loginPresenter.setCredentialError('message');

    fixture.detectChanges();

    expect(fixture.nativeElement).toMatchSnapshot();
  });
});
