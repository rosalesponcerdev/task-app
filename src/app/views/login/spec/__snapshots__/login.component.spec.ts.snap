// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginComponent Loggin Error 1`] = `
<div
  id="root2"
>
  <div
    class="container"
  >
    <div
      class="row align-items-center vh-100"
    >
      <div
        class="col"
      >
        <div
          class="w-50 mx-auto"
        >
          <fieldset
            class="border border-2 px-3 pt-3 pb-5"
          >
            <legend
              class="float-none m-0"
            >
              LOGIN
            </legend>
            <span
              class="text-danger fw-bold"
            >
               | 
            </span>
            <form
              class="row align-items-center ng-untouched ng-pristine ng-invalid"
              novalidate=""
            >
              <div
                class="col-8"
              >
                <div
                  class="mb-3"
                >
                  <label
                    class="form-label"
                    for="usernameinput"
                  >
                    Usuario
                  </label>
                  <input
                    autocomplete="off"
                    class="form-control ng-untouched ng-pristine ng-invalid"
                    formcontrolname="username"
                    id="usernameinput"
                    type="text"
                  />
                </div>
                <div>
                  <label
                    class="form-label"
                    for="passinput"
                  >
                    Contrase単a
                  </label>
                  <input
                    class="form-control ng-untouched ng-pristine ng-invalid"
                    formcontrolname="password"
                    id="passinput"
                    type="password"
                  />
                </div>
              </div>
              <div
                class="col-4 text-center"
              >
                <button
                  class="btn btn-primary"
                  disabled=""
                  type="submit"
                >
                  <span>
                    Ingresar
                  </span>
                  
                  
                </button>
              </div>
            </form>
          </fieldset>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`LoginComponent Loggin Error 2`] = `
<div
  id="root2"
>
  <div
    class="container"
  >
    <div
      class="row align-items-center vh-100"
    >
      <div
        class="col"
      >
        <div
          class="w-50 mx-auto"
        >
          <fieldset
            class="border border-2 px-3 pt-3 pb-5"
          >
            <legend
              class="float-none m-0"
            >
              LOGIN
            </legend>
            <span
              class="text-danger fw-bold"
            >
               | 
            </span>
            <form
              class="row align-items-center ng-untouched ng-pristine ng-invalid"
              novalidate=""
            >
              <div
                class="col-8"
              >
                <div
                  class="mb-3"
                >
                  <label
                    class="form-label"
                    for="usernameinput"
                  >
                    Usuario
                  </label>
                  <input
                    autocomplete="off"
                    class="form-control ng-untouched ng-pristine ng-invalid"
                    formcontrolname="username"
                    id="usernameinput"
                    type="text"
                  />
                </div>
                <div>
                  <label
                    class="form-label"
                    for="passinput"
                  >
                    Contrase単a
                  </label>
                  <input
                    class="form-control ng-untouched ng-pristine ng-invalid"
                    formcontrolname="password"
                    id="passinput"
                    type="password"
                  />
                </div>
              </div>
              <div
                class="col-4 text-center"
              >
                <button
                  class="btn btn-primary"
                  disabled=""
                  type="submit"
                >
                  <span>
                    Ingresar
                  </span>
                  
                  
                </button>
              </div>
            </form>
          </fieldset>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`LoginComponent Loggin Success 1`] = `
<div
  id="root1"
>
  <div
    class="container"
  >
    <div
      class="row align-items-center vh-100"
    >
      <div
        class="col"
      >
        <div
          class="w-50 mx-auto"
        >
          <fieldset
            class="border border-2 px-3 pt-3 pb-5"
          >
            <legend
              class="float-none m-0"
            >
              LOGIN
            </legend>
            <span
              class="text-danger fw-bold"
            >
               | 
            </span>
            <form
              class="row align-items-center ng-untouched ng-pristine ng-invalid"
              novalidate=""
            >
              <div
                class="col-8"
              >
                <div
                  class="mb-3"
                >
                  <label
                    class="form-label"
                    for="usernameinput"
                  >
                    Usuario
                  </label>
                  <input
                    autocomplete="off"
                    class="form-control ng-untouched ng-pristine ng-invalid"
                    formcontrolname="username"
                    id="usernameinput"
                    type="text"
                  />
                </div>
                <div>
                  <label
                    class="form-label"
                    for="passinput"
                  >
                    Contrase単a
                  </label>
                  <input
                    class="form-control ng-untouched ng-pristine ng-invalid"
                    formcontrolname="password"
                    id="passinput"
                    type="password"
                  />
                </div>
              </div>
              <div
                class="col-4 text-center"
              >
                <button
                  class="btn btn-primary"
                  disabled=""
                  type="submit"
                >
                  <span>
                    Ingresar
                  </span>
                  
                  
                </button>
              </div>
            </form>
          </fieldset>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`LoginComponent Loggin Success 2`] = `
<div
  id="root1"
>
  <div
    class="container"
  >
    <div
      class="row align-items-center vh-100"
    >
      <div
        class="col"
      >
        <div
          class="w-50 mx-auto"
        >
          <fieldset
            class="border border-2 px-3 pt-3 pb-5"
          >
            <legend
              class="float-none m-0"
            >
              LOGIN
            </legend>
            <span
              class="text-danger fw-bold"
            >
               | 
            </span>
            <form
              class="row align-items-center ng-untouched ng-pristine ng-invalid"
              novalidate=""
            >
              <div
                class="col-8"
              >
                <div
                  class="mb-3"
                >
                  <label
                    class="form-label"
                    for="usernameinput"
                  >
                    Usuario
                  </label>
                  <input
                    autocomplete="off"
                    class="form-control ng-untouched ng-pristine ng-invalid"
                    formcontrolname="username"
                    id="usernameinput"
                    type="text"
                  />
                </div>
                <div>
                  <label
                    class="form-label"
                    for="passinput"
                  >
                    Contrase単a
                  </label>
                  <input
                    class="form-control ng-untouched ng-pristine ng-invalid"
                    formcontrolname="password"
                    id="passinput"
                    type="password"
                  />
                </div>
              </div>
              <div
                class="col-4 text-center"
              >
                <button
                  class="btn btn-primary"
                  disabled=""
                  type="submit"
                >
                  <span>
                    Ingresar
                  </span>
                  
                  
                </button>
              </div>
            </form>
          </fieldset>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`LoginComponent should create 1`] = `
<div
  id="root0"
>
  <div
    class="container"
  >
    <div
      class="row align-items-center vh-100"
    >
      <div
        class="col"
      >
        <div
          class="w-50 mx-auto"
        >
          <fieldset
            class="border border-2 px-3 pt-3 pb-5"
          >
            <legend
              class="float-none m-0"
            >
              LOGIN
            </legend>
            <span
              class="text-danger fw-bold"
            >
               | 
            </span>
            <form
              class="row align-items-center ng-untouched ng-pristine ng-invalid"
              novalidate=""
            >
              <div
                class="col-8"
              >
                <div
                  class="mb-3"
                >
                  <label
                    class="form-label"
                    for="usernameinput"
                  >
                    Usuario
                  </label>
                  <input
                    autocomplete="off"
                    class="form-control ng-untouched ng-pristine ng-invalid"
                    formcontrolname="username"
                    id="usernameinput"
                    type="text"
                  />
                </div>
                <div>
                  <label
                    class="form-label"
                    for="passinput"
                  >
                    Contrase単a
                  </label>
                  <input
                    class="form-control ng-untouched ng-pristine ng-invalid"
                    formcontrolname="password"
                    id="passinput"
                    type="password"
                  />
                </div>
              </div>
              <div
                class="col-4 text-center"
              >
                <button
                  class="btn btn-primary"
                  disabled=""
                  type="submit"
                >
                  <span>
                    Ingresar
                  </span>
                  
                  
                </button>
              </div>
            </form>
          </fieldset>
        </div>
      </div>
    </div>
  </div>
</div>
`;
